src/main/java/org/opensearch/knn/index/query/nativelib/NativeEngineKnnVectorQuery.java

```diff
@@ -76,16 +77,71 @@ public Weight createWeight(IndexSearcher indexSearcher, ScoreMode scoreMode, flo
             log.debug("Rescoring results took {} ms. oversampled k:{}, segments:{}", rescoreTime, firstPassK, leafReaderContexts.size());
         }
         ResultUtil.reduceToTopK(perLeafResults, finalK);
+
+        if (isInnerHitQuery) {
+            perLeafResults = retrieveAll(indexSearcher, leafReaderContexts, knnWeight, perLeafResults);
+        }
+
         TopDocs[] topDocs = new TopDocs[perLeafResults.size()];
         for (int i = 0; i < perLeafResults.size(); i++) {
             topDocs[i] = ResultUtil.resultMapToTopDocs(perLeafResults.get(i), leafReaderContexts.get(i).docBase);
         }
 
-        TopDocs topK = TopDocs.merge(knnQuery.getK(), topDocs);
+        TopDocs topK = TopDocs.merge(getTotalTopDoc(topDocs), topDocs);
+
         if (topK.scoreDocs.length == 0) {
             return new MatchNoDocsQuery().createWeight(indexSearcher, scoreMode, boost);
         }
-        return createDocAndScoreQuery(reader, topK).createWeight(indexSearcher, scoreMode, boost);
+        return queryUtils.createDocAndScoreQuery(reader, topK).createWeight(indexSearcher, scoreMode, boost);
+    }
+
+    private int getTotalTopDoc(TopDocs[] topDocs) {
+        if (isInnerHitQuery == false) {
+            return knnQuery.getK();
+        }
+
+        int sum = 0;
+        for (TopDocs topDoc : topDocs) {
+            sum += topDoc.scoreDocs.length;
+        }
+        return sum;
+    }
+
+    private List<Map<Integer, Float>> retrieveAll(
+        final IndexSearcher indexSearcher,
+        List<LeafReaderContext> leafReaderContexts,
+        KNNWeight knnWeight,
+        List<Map<Integer, Float>> perLeafResults
+    ) throws IOException {
+        List<Callable<Map<Integer, Float>>> nestedQueryTasks = new ArrayList<>(leafReaderContexts.size());
+        for (int i = 0; i < perLeafResults.size(); i++) {
+            LeafReaderContext leafReaderContext = leafReaderContexts.get(i);
+            int finalI = i;
+            nestedQueryTasks.add(() -> {
+                Bits queryFilter = queryUtils.createBits(leafReaderContext, knnWeight.getFilterWeight());
```

Comment:
Lets create the Bits for filter in NativeEngineKnnVectorQuery, there is another case for this where the [bits](https://github.com/opensearch-project/k-NN/pull/2273#discussion_r1850774606) are needed, it makes sense to compute once and pass it around

Reply:
We can refactor when it is really needed. If the PR you referenced is merged first, I will try to reuse the Bits.